### PR TITLE
Support Map and Array Type in New Parquet Reader

### PR DESCRIPTION
**Pull request by zhenxiao**
@nezihyigitbasi @dain could you please review when you are free?

### CODE REVIEW

**Comment by zhenxiao** [comment]
@nezihyigitbasi @dain kindly ping

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 1]
```diff
@@ -124,7 +133,57 @@ private boolean advanceToNextRowGroup()
...
+
+        if (lengths.size() == 0) {
+            return RunLengthEncodedBlock.create(parameters.get(0), null, batchSize);
+        }
+
+        int[] offsets = new int[batchSize + 1];
+        for (int i = 1; i < offsets.length; i++) {
+            offsets[i] = offsets[i - 1] + lengths.get(i - 1);
+        }
+        return new ArrayBlock(batchSize, new boolean[batchSize], offsets, block);
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: Two comments here regarding the `isNull` array passed to the `ArrayBlock` constructor:
- Why don't we populate this array properly instead of passing a brand new one with each call? Are we sure that there is no other method that depends on this array to be valid?
- We are creating these `1K` `isNull` boolean arrays and throw them away with each call, although they are small when we have a large number of rows this will produce some garbage on the heap. If we really don't need them we can have a static final boolean array that's init once instead of creating a new one all the time.

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 2]
```diff
@@ -195,4 +246,76 @@ private void initializeColumnReaders()
             columnReadersMap.put(column, ParquetColumnReader.createReader(column));
         }
     }
+
+    private Block readBlock(String name, Type type, List<String> path, List<Integer> offsets, List<Integer> positions)
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: `offsets` is called `lengths` in some parts of the code (e.g., in `readMap` and `readArray`), we should stick to one of these (offset is better I think).

**Comment by dain** [comment]
@zhenxiao, when you get @nezihyigitbasi's changes resolved, reassign to me

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 3]
```diff
@@ -194,6 +177,76 @@ public Block readPrimitive(Type type)
...
+                    if (repetitionLevel == 0) {
+                        lengths.add(valueCount);
+                        valueCount = 0;
+                        if (i == valueNumber - 1) {
+                            return;
+                        }
+                    }
+                    definitionLevel = definitionReader.readLevel();
+                }
+                catch (Exception e) {
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: repetitionReader might throw an exception when it hits the end of repetition streams

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java [hunk 4]
```diff
@@ -130,7 +129,7 @@ public ParquetPageSourceFactory(TypeManager typeManager, boolean useParquetColum
             return Optional.empty();
         }
 
-        return Optional.of(createParquetPageSource(
+        return createParquetPageSource(
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: get it, will update

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java [hunk 5]
```diff
@@ -235,7 +234,8 @@ private static boolean columnTypeSupported(List<HiveColumnHandle> columns)
         boolean regularColumnsSupported = columns.stream()
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: thanks

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetBinaryColumnReader.java [hunk 6]
```diff
@@ -41,39 +41,35 @@ public BlockBuilder createBlockBuilder(Type type)
...
-                blockBuilder.appendNull();
+                value = wrappedBuffer(binary.getBytes());
+            }
+            if (isVarcharType(type)) {
+                value = truncateToLength(value, type);
+            }
+            if (isCharType(type)) {
+                value = trimSpacesAndTruncateToLength(value, type);
             }
+                type.writeSlice(blockBuilder, value);
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetBinaryColumnReader.java]: wil update

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 7]
```diff
@@ -35,6 +35,7 @@
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: get it, thanks

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 8]
```diff
@@ -50,24 +51,22 @@
...
     private ParquetPageReader pageReader;
     private ParquetDictionary dictionary;
-    private int repetitionLevel;
-    private int definitionLevel;
     private int currentValueCount;
     private ParquetDataPage page;
     private int remainingValueCountInPage;
     private int readOffset;
 
-    public abstract BlockBuilder createBlockBuilder(Type type);
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: just would like to protect so that other classes are not calling these methods

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 9]
```diff
@@ -156,34 +155,18 @@ public long getTotalValueCount()
         return totalValueCount;
     }
 
-    public Block readPrimitive(Type type)
+    public Block readPrimitive(Type type, List<Integer> lengths)
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: get it, thanks

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 10]
```diff
@@ -194,6 +177,76 @@ public Block readPrimitive(Type type)
         return blockBuilder.build();
     }
 
+    private void readGroup(BlockBuilder blockBuilder, int valueNumber, Type type, List<Integer> lengths)
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: updated to numValues

Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: get it, thanks

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 11]
```diff
@@ -194,6 +177,76 @@ public Block readPrimitive(Type type)
...
+                }
+                catch (Exception e) {
+                    lengths.add(valueCount);
+                    return;
+                }
+            } while (repetitionLevel != 0);
+        }
+    }
+
+    private void skipGroup(int offset)
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: get it, thanks

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 12]
```diff
@@ -194,6 +177,76 @@ public Block readPrimitive(Type type)
...
+    {
+        definitionLevel = definitionReader.readLevel();
+        repetitionLevel = repetitionReader.readLevel();
+        for (int i = 0; i < offset; i++) {
+            do {
+                skipValues();
+                try {
+                    repetitionLevel = repetitionReader.readLevel();
+                    if (i == offset - 1 && repetitionLevel == 0) {
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: would like to break the loop when it is a start of new list, so keep it as AND

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 13]
```diff
@@ -194,6 +177,76 @@ public Block readPrimitive(Type type)
...
+            } while (repetitionLevel != 0);
+        }
+    }
+
+    private void seek()
+            throws IOException
+    {
+        checkArgument(currentValueCount <= totalValueCount, "Already read all values in column chunk");
+        // Parquet does not have api to skip in datastream, have to skip values
+        // TODO skip in datastream
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: yep, mainly need it from the parquet side, let me remove it

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 14]
```diff
@@ -194,6 +177,76 @@ public Block readPrimitive(Type type)
...
+        if (readOffset != 0) {
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: thanks, will update

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 15]
```diff
@@ -194,6 +177,76 @@ public Block readPrimitive(Type type)
...
+            while (valuePosition < readOffset) {
+                if (page == null) {
+                    readNextPage();
+                }
+                int offset = Math.min(remainingValueCountInPage, readOffset - valuePosition);
+                skipGroup(offset);
+                valuePosition = valuePosition + offset;
+                updatePosition(offset);
+            }
+            checkArgument(valuePosition == readOffset, "valuePosition " + valuePosition + " not equals to readOffset " + readOffset);
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: get it, will update

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 16]
```diff
@@ -124,7 +133,57 @@ private boolean advanceToNextRowGroup()
         return true;
     }
 
-    public Block readStruct(Type type, List<String> path)
+    public Block readArray(Type type, List<String> path, List<Integer> positions)
+            throws IOException
+    {
+        List<Type> parameters = type.getTypeParameters();
+        checkArgument(parameters.size() == 1, "Array must has 1 parameter, this array has %d parameters", parameters.size());
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: get it, will update, thank u

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 17]
```diff
@@ -124,7 +133,57 @@ private boolean advanceToNextRowGroup()
...
+            offsets[i] = offsets[i - 1] + lengths.get(i - 1);
+        }
+        return new ArrayBlock(batchSize, new boolean[batchSize], offsets, block);
+    }
+
+    public Block readMap(Type type, List<String> path, List<Integer> positions)
+            throws IOException
+    {
+        List<Type> parameters = type.getTypeParameters();
+        checkArgument(parameters.size() == 2, "Map must has 2 parameters, this map has %d parameters", parameters.size());
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: get it, will update

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 18]
```diff
@@ -133,32 +192,20 @@ public Block readStruct(Type type, List<String> path)
...
-            }
-
-            if (ROW.equals(fieldType.getTypeSignature().getBase())) {
-                blocks[i] = readStruct(fieldType, path);
-            }
-            else {
-                blocks[i] = readPrimitive(columnDescriptor.get(), fieldType);
-            }
-            path.remove(name);
+            List<Integer> lengths = new ArrayList<>();
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: get it, will update

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java (hunk 2 above)
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: get it, will update

**Comment by zhenxiao** [comment]
resolve @nezihyigitbasi 's comments, thank you @nezihyigitbasi

**Comment by zhenxiao** [comment]
@dain your comments and suggestions are appreciated

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 19]
```diff
@@ -136,62 +135,100 @@ public void prepareNextRead(int batchSize)
...
     {
-        return columnDescriptor;
+        definitionLevel = definitionReader.readLevel();
+        repetitionLevel = repetitionReader.readLevel();
+        int valueCount = 0;
+        for (int i = 0; i < numValues; i++) {
+            do {
+                readValue(blockBuilder, type);
+                try {
+                    valueCount = valueCount + 1;
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: use `++`

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java (hunk 3 above)
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: so  it is an "expected" exception that you want to ignore?  If so, renamed `e` to `expected`.  Also, can the catch type be narrower?  If not, you will need to handle `InterruptedException` and `PrestoException` specially. Also, add a comment explaining exactly why this happens and why it is acceptable.

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 20]
```diff
@@ -136,62 +135,100 @@ public void prepareNextRead(int batchSize)
...
+            do {
+                skipValue();
+                try {
+                    repetitionLevel = repetitionReader.readLevel();
+                    if (i == offset - 1 && repetitionLevel == 0) {
+                        return;
+                    }
+                    definitionLevel = definitionReader.readLevel();
+                }
+                catch (Exception e) {
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: same here

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 21]
```diff
@@ -124,7 +133,57 @@ private boolean advanceToNextRowGroup()
...
 
-    public Block readStruct(Type type, List<String> path)
+    public Block readArray(Type type, List<String> path, List<Integer> positions)
+            throws IOException
+    {
+        List<Type> parameters = type.getTypeParameters();
+        checkArgument(parameters.size() == 1, "Arrays must have a single type parameter, found %d", parameters.size());
+        List<Integer> elementOffsets = new ArrayList<>();
+        path.add(ARRAY_TYPE_NAME);
+        Block block = readBlock(ARRAY_ELEMENT_NAME, parameters.get(0), path, elementOffsets, positions);
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: assign `parameters.get(0)` to a variable named `elementType`

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 22]
```diff
@@ -124,7 +133,57 @@ private boolean advanceToNextRowGroup()
...
+        path.remove(ARRAY_TYPE_NAME);
+
+        if (elementOffsets.size() == 0) {
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: `if (elementOffsets.isEmpty()) {`  IntelliJ should warn you about this

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 23]
```diff
@@ -195,4 +245,76 @@ private void initializeColumnReaders()
...
+            getBlockOffsets(block, offsets);
+        }
+        else {
+            block = readPrimitive(descriptor.get(), type, offsets, positions);
+        }
+        path.remove(name);
+        return block;
+    }
+
+    private void getBlockOffsets(Block block, List<Integer> offsets)
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: the next three methods should be static (again IntelliJ should warn you)

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 24]
```diff
@@ -195,4 +245,76 @@ private void initializeColumnReaders()
...
+    {
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            offsets.add(block.getLength(i));
+        }
+    }
+
+    private List<Integer> getNonNullPositions(int positionCount, Block keys, List<Integer> offsets)
+    {
+        List<Integer> positions = new ArrayList(keys.getPositionCount());
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: missing diamond ` List<Integer> positions = new ArrayList<>(keys.getPositionCount());` (again IntelliJ should warn you)

File: presto-hive/src/test/java/com/facebook/presto/hive/parquet/ParquetTester.java [hunk 25]
```diff
@@ -182,7 +183,7 @@ private static void assertFileContents(JobConf jobConf,
         Iterator<?> iterator = expectedValues.iterator();
         for (int batchSize = parquetReader.nextBatch(); batchSize >= 0; batchSize = parquetReader.nextBatch()) {
             ColumnDescriptor columnDescriptor = fileSchema.getColumns().get(0);
-            Block block = parquetReader.readPrimitive(columnDescriptor, type);
+            Block block = parquetReader.readPrimitive(columnDescriptor, type, new ArrayList<>(), new ArrayList<>());
```
Comment (dain) [inline presto-hive/src/test/java/com/facebook/presto/hive/parquet/ParquetTester.java]: Can you port over the struct, list and map tests from the ORC or RCFile tester code?

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java (hunk 2 above)
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: If you use `it.unimi.dsi.fastutil.ints.IntList` with the `getInt` style methods, you will get better performance

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 26]
```diff
@@ -133,32 +192,19 @@ public Block readStruct(Type type, List<String> path)
...
+        int[] offsets = new int[blockSize + 1];
         for (int i = 1; i < offsets.length; i++) {
             offsets[i] = i * parameters.size();
         }
-        return new ArrayBlock(batchSize, new boolean[batchSize], offsets, interleavedBlock);
+        return new ArrayBlock(blockSize, new boolean[blockSize], offsets, interleavedBlock);
     }
 
-    public Block readPrimitive(ColumnDescriptor columnDescriptor, Type type)
+    public Block readPrimitive(ColumnDescriptor columnDescriptor, Type type, List<Integer> offsets, List<Integer> positions)
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: There is something strange with this method name.  `ParquetBlockLoader` calls this for all reads, so I assume this works for all types.

Additionally, I would make this `protected` and add a `public` variant that does not take the list arguments

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 27]
```diff
@@ -173,7 +219,11 @@ public Block readPrimitive(ColumnDescriptor columnDescriptor, Type type)
             ParquetColumnChunk columnChunk = new ParquetColumnChunk(descriptor, buffer, 0);
             columnReader.setPageReader(columnChunk.readAllPages());
         }
-        return columnReader.readPrimitive(type);
+        Block block = columnReader.readPrimitive(type, offsets);
+        if (positions.isEmpty()) {
+            return block;
+        }
+        return block.copyPositions(positions);
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: I'm having a hard time understand this code.  I don't really understand how `offsets` and `positions` are used, and I really don't understand why you would have copyPositions, which is super expensive, here.

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java (hunk 3 above)
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: get it. will update to specific exceptions

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java (hunk 2 above)
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: get it. will update

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 28]
```diff
@@ -195,4 +246,76 @@ private void initializeColumnReaders()
...
+        Optional<RichColumnDescriptor> descriptor = getDescriptor(fileSchema, requestedSchema, path);
+        if (!descriptor.isPresent()) {
+            path.remove(name);
+            if (positions.isEmpty()) {
+                return RunLengthEncodedBlock.create(type, null, batchSize);
+            }
+            return RunLengthEncodedBlock.create(type, null, positions.size());
+        }
+
+        Block block = null;
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: get it, will update

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java [hunk 29]
```diff
@@ -86,7 +85,7 @@
             .add("parquet.hive.serde.ParquetHiveSerDe")
             .build();
     @VisibleForTesting
-    public static final Set<String> SUPPORTED_COLUMN_TYPES = ImmutableSet.of(INTEGER, BIGINT, BOOLEAN, DOUBLE, REAL, TIMESTAMP, VARCHAR, CHAR, VARBINARY, DATE, DECIMAL, ROW);
+    public static final Set<String> SUPPORTED_COLUMN_TYPES = ImmutableSet.of(INTEGER, BIGINT, BOOLEAN, DOUBLE, REAL, TIMESTAMP, VARCHAR, CHAR, VARBINARY, DATE, DECIMAL, ROW, MAP, ARRAY);
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: all supported now. Will remove this

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 30]
```diff
@@ -50,24 +51,22 @@
 
     protected ParquetLevelReader repetitionReader;
     protected ParquetLevelReader definitionReader;
+    protected int repetitionLevel;
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: get it, thanks

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 31]
```diff
@@ -50,24 +51,22 @@
...
     private int readOffset;
 
-    public abstract BlockBuilder createBlockBuilder(Type type);
-
-    public abstract void readValues(BlockBuilder blockBuilder, int valueNumber, Type type);
-
-    public abstract void skipValues(int offsetNumber);
+    protected abstract BlockBuilder createBlockBuilder(Type type);
+    protected abstract void readValue(BlockBuilder blockBuilder, Type type);
+    protected abstract void skipValue();
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: get it, will update

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 32]
```diff
@@ -136,62 +135,100 @@ public void prepareNextRead(int batchSize)
...
+        return columnDescriptor;
     }
 
-    public int getCurrentDefinitionLevel()
+    public Block readPrimitive(Type type, List<Integer> positions)
+            throws IOException
     {
-        return definitionLevel;
+        seek();
+        BlockBuilder blockBuilder = createBlockBuilder(type);
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: get it, thank you

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 33]
```diff
@@ -136,62 +135,100 @@ public void prepareNextRead(int batchSize)
...
+        int valueCount = 0;
+        while (valueCount < nextBatchSize) {
+            if (page == null) {
+                readNextPage();
+            }
+            int numValues = Math.min(remainingValueCountInPage, nextBatchSize - valueCount);
+            readValues(blockBuilder, numValues, type, positions);
+            valueCount = valueCount + numValues;
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: get it, thanks

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java (hunk 19 above)
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: get it, thanks

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java (hunk 20 above)
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: get it, will update

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java (hunk 21 above)
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: will update

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java (hunk 22 above)
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: will update

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java (hunk 26 above)
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: readPrimitive() is just for primitive, Map, Array, Struct will go to readMap, readArray, readStruct, and finally call readPrimitive()

Will add public variant and make it protected

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java (hunk 27 above)
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: trying to get rid of null keys for maps. Let me try rebase it

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java (hunk 23 above)
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: get it, will update

**Comment by zhenxiao** [comment]
get comments addressed, thank you @dain 
I will port ORC/RC tests for Array/Map/Struct in a following commit

**Comment by zhenxiao** [comment]
kindly ping

**Comment by dain** [comment]
The check in `ParquetPageSourceFactory` for supported partition column types is no longer needed, since Hive readers don't see partition columns anymore (they are handled automatically for readers now).

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java [hunk 34]
```diff
@@ -136,62 +136,102 @@ public void prepareNextRead(int batchSize)
...
+                        positions.add(valueCount);
+                        valueCount = 0;
+                        if (i == numValues - 1) {
+                            return;
+                        }
+                    }
+                    definitionLevel = definitionReader.readLevel();
+                }
+                catch (IllegalArgumentException expected) {
+                // Reading past repetition stream, RunLengthBitPackingHybridDecoder throws IllegalArgumentException
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetColumnReader.java]: formatting

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 35]
```diff
@@ -133,34 +190,27 @@ public Block readStruct(Type type, List<String> path)
...
+        return new ArrayBlock(blockSize, new boolean[blockSize], offsets, interleavedBlock);
     }
 
     public Block readPrimitive(ColumnDescriptor columnDescriptor, Type type)
             throws IOException
     {
+        return readPrimitive(columnDescriptor, type, new IntArrayList());
+    }
+
+    protected Block readPrimitive(ColumnDescriptor columnDescriptor, Type type, IntList offsets)
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: this can be `private`

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java [hunk 36]
```diff
@@ -186,11 +188,18 @@ public Page getNextPage()
                         continue;
                     }
 
+                    String fieldName = fileSchema.getFields().get(fieldIndex).getName();
                     List<String> path = new ArrayList<>();
-                    path.add(fileSchema.getFields().get(fieldIndex).getName());
+                    path.add(fieldName);
                     if (ROW.equals(type.getTypeSignature().getBase())) {
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java]: All of these should use `LazyBlock`

Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java]: Also why don't these have the `getDescriptor(fileSchema, requestedSchema, path).isPresent()` check to determine if the column is in the file?

Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java]: I think we should move this `if/else` block inside of `ParquetReader` and change its interface to have only a `readBlock (ColumnDescriptor, Type)` method.

Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java]: You can make these changes in another commit or PR if you want.

**Comment by dain** [comment]
@nezihyigitbasi can you "Approve changes" on your comment or add new requests?

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 37]
```diff
@@ -195,4 +245,41 @@ private void initializeColumnReaders()
             columnReadersMap.put(column, ParquetColumnReader.createReader(column));
         }
     }
+
+    private Block readBlock(String name, Type type, List<String> path, IntList offsets)
+            throws IOException
+    {
+        path.add(name);
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: any reason why you don't make this `path` variable (and the ones in other methods) immutable so that you don't need to do add/remove? I see similar add/remove logic around. And I am assuming these methods are accessed from a single thread, right?

**Comment by nezihyigitbasi** [comment]
LGTM % my single comment about the `path` variable.

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java (hunk 36 above)
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java]: get it. Thank you @dain I will submit another PR to resolve this

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java (hunk 37 above)
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: @nezihyigitbasi I am using path variable to recursively process nested path names. Need to add new path components when getting into a deeper level, and remove it when deeper level is done, comes back to the current level. Yes, it is a single thread. Let me see whether I could make it more clear

**Comment by zhenxiao** [comment]
@dain @nezihyigitbasi thank you so much, comments addressed with an updated version

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java (hunk 36 above)
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java]: Can you file an issue for this and the "port ORC tests", and assign them to yourself, so we don't loose track of them?

**Comment by dain** [comment]
I'll merge this after the current release goes out.

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java (hunk 36 above)
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java]: yeah:
https://github.com/prestodb/presto/issues/7187
https://github.com/prestodb/presto/issues/7188

**Comment by dain** [comment]
Merged, thanks!
